File: translate_all_languages.py
#!/usr/bin/env python3
"""
Translation Script for MicronsHub
Translates the complete German translation file to all other languages
"""

import json
import os
from googletrans import Translator
import time

# Language codes mapping
LANGUAGES = {
    'fr': 'French',
    'es': 'Spanish', 
    'it': 'Italian',
    'nl': 'Dutch',
    'pl': 'Polish',
    'pt': 'Portuguese',
    'sv': 'Swedish',
    'da': 'Danish',
    'fi': 'Finnish',
    'nb': 'Norwegian',
    'hu': 'Hungarian',
    'cs': 'Czech'
}

def load_german_translations():
    """Load the complete German translation file"""
    try:
        with open('src/locales/de/translation.json', 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        print("❌ German translation file not found!")
        return None
    except json.JSONDecodeError:
        print("❌ Invalid JSON in German translation file!")
        return None

def translate_text(text, target_lang, translator):
    """Translate a single text to target language"""
    try:
        # Add delay to avoid rate limiting
        time.sleep(0.1)
        result = translator.translate(text, dest=target_lang)
        return result.text
    except Exception as e:
        print(f"⚠️  Translation error for '{text[:50]}...': {e}")
        return text  # Return original text if translation fails

def translate_file(german_data, target_lang, translator):
    """Translate the entire German file to target language"""
    print(f"🔄 Translating to {LANGUAGES[target_lang]} ({target_lang})...")
    
    translated_data = {}
    total_keys = len(german_data.keys())
    
    for i, (key, value) in enumerate(german_data.items(), 1):
        if isinstance(value, str):
            # Translate string values
            translated_value = translate_text(value, target_lang, translator)
            translated_data[key] = translated_value
        elif isinstance(value, list):
            # Translate list values
            translated_list = []
            for item in value:
                if isinstance(item, str):
                    translated_item = translate_text(item, target_lang, translator)
                    translated_list.append(translated_item)
                else:
                    translated_list.append(item)
            translated_data[key] = translated_list
        else:
            # Keep non-string values as-is (numbers, booleans, etc.)
            translated_data[key] = value
        
        # Progress indicator
        if i % 50 == 0:
            print(f"   Progress: {i}/{total_keys} keys translated")
    
    return translated_data

def save_translation(translated_data, target_lang):
    """Save translated data to file"""
    output_file = f'src/locales/{target_lang}/translation.json'
    
    # Ensure directory exists
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    
    # Save with proper formatting
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(translated_data, f, ensure_ascii=False, indent=2)
    
    print(f"✅ Saved {target_lang} translation ({len(translated_data)} keys)")

def main():
    """Main translation process"""
    print("🚀 Starting translation of German file to all languages...")
    
    # Load German translations
    german_data = load_german_translations()
    if not german_data:
        return
    
    print(f"📖 Loaded German file with {len(german_data)} translation keys")
    
    # Initialize translator
    translator = Translator()
    
    # Translate to each language
    for lang_code in LANGUAGES.keys():
        try:
            # Skip German (source language)
            if lang_code == 'de':
                continue
                
            # Translate the file
            translated_data = translate_file(german_data, lang_code, translator)
            
            # Save the translation
            save_translation(translated_data, lang_code)
            
            print(f"✅ Completed {LANGUAGES[lang_code]} translation")
            print("-" * 50)
            
        except Exception as e:
            print(f"❌ Error translating to {lang_code}: {e}")
            continue
    
    print("🎉 Translation process completed!")
    print("\n📊 Summary:")
    print(f"- Source: German ({len(german_data)} keys)")
    print(f"- Translated to: {len(LANGUAGES)} languages")
    print("- Languages: " + ", ".join(LANGUAGES.keys()))

if __name__ == "__main__":
    main() 

File: scripts/translate_all_languages.js
#!/usr/bin/env node
/**
 * Translation Script for MicronsHub
 * Translates the complete German translation file to all other languages
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Language codes mapping
const LANGUAGES = {
    'fr': 'French',
    'es': 'Spanish', 
    'it': 'Italian',
    'nl': 'Dutch',
    'pl': 'Polish',
    'pt': 'Portuguese',
    'sv': 'Swedish',
    'da': 'Danish',
    'fi': 'Finnish',
    'nb': 'Norwegian',
    'hu': 'Hungarian',
    'cs': 'Czech'
};

// Simple translation function using Google Translate API
async function translateText(text, targetLang) {
    try {
        // Using Google Translate API endpoint
        const response = await fetch(`https://translate.googleapis.com/translate_a/single?client=gtx&sl=de&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}`);
        const data = await response.json();
        return data[0][0][0];
    } catch (error) {
        console.log(`⚠️  Translation error for '${text.substring(0, 50)}...': ${error.message}`);
        return text; // Return original text if translation fails
    }
}

function loadGermanTranslations() {
    try {
        const filePath = path.join(__dirname, '..', 'src', 'locales', 'de', 'translation.json');
        const data = fs.readFileSync(filePath, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        console.log('❌ German translation file not found or invalid!');
        return null;
    }
}

async function translateFile(germanData, targetLang) {
    console.log(`🔄 Translating to ${LANGUAGES[targetLang]} (${targetLang})...`);
    
    const translatedData = {};
    const totalKeys = Object.keys(germanData).length;
    
    for (let i = 0; i < totalKeys; i++) {
        const key = Object.keys(germanData)[i];
        const value = germanData[key];
        
        if (typeof value === 'string') {
            // Translate string values
            const translatedValue = await translateText(value, targetLang);
            translatedData[key] = translatedValue;
        } else if (Array.isArray(value)) {
            // Translate list values
            const translatedList = [];
            for (const item of value) {
                if (typeof item === 'string') {
                    const translatedItem = await translateText(item, targetLang);
                    translatedList.push(translatedItem);
                } else {
                    translatedList.push(item);
                }
            }
            translatedData[key] = translatedList;
        } else {
            // Keep non-string values as-is (numbers, booleans, etc.)
            translatedData[key] = value;
        }
        
        // Progress indicator
        if ((i + 1) % 50 === 0) {
            console.log(`   Progress: ${i + 1}/${totalKeys} keys translated`);
        }
        
        // Add delay to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    return translatedData;
}

function saveTranslation(translatedData, targetLang) {
    const outputDir = path.join(__dirname, '..', 'src', 'locales', targetLang);
    const outputFile = path.join(outputDir, 'translation.json');
    
    // Ensure directory exists
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    
    // Save with proper formatting
    fs.writeFileSync(outputFile, JSON.stringify(translatedData, null, 2), 'utf8');
    
    console.log(`✅ Saved ${targetLang} translation (${Object.keys(translatedData).length} keys)`);
}

async function main() {
    console.log('🚀 Starting translation of German file to all languages...');
    
    // Load German translations
    const germanData = loadGermanTranslations();
    if (!germanData) {
        return;
    }
    
    console.log(`📖 Loaded German file with ${Object.keys(germanData).length} translation keys`);
    
    // Translate to each language
    for (const langCode of Object.keys(LANGUAGES)) {
        try {
            // Skip German (source language)
            if (langCode === 'de') {
                continue;
            }
            
            // Translate the file
            const translatedData = await translateFile(germanData, langCode);
            
            // Save the translation
            saveTranslation(translatedData, langCode);
            
            console.log(`✅ Completed ${LANGUAGES[langCode]} translation`);
            console.log('-'.repeat(50));
            
        } catch (error) {
            console.log(`❌ Error translating to ${langCode}: ${error.message}`);
            continue;
        }
    }
    
    console.log('🎉 Translation process completed!');
    console.log('\n📊 Summary:');
    console.log(`- Source: German (${Object.keys(germanData).length} keys)`);
    console.log(`- Translated to: ${Object.keys(LANGUAGES).length} languages`);
    console.log(`- Languages: ${Object.keys(LANGUAGES).join(', ')}`);
}

// Run the script
main().catch(console.error); 